### PR TITLE
Allow keyPrefix to accept an additional argument for double maps

### LINKED ISSUE #2213: get key prefix for a double map with first key argument
### Goal:

Update `keyPrefix` for double map storage to take one argument as the first key and return the key prefix.

### Background

Currently polkadot.js provides two methods to work with the storage key value:

`keyPrefix()` that return the storage prefix.
`key(args)` that return the whole storage key.

None of it able to provide key prefix for a double map with a given first key item.

### Actions:

Modify `keyPrefix`, change signature to `(key1?: CodecArg): string`

https://github.com/polkadot-js/api/blob/e77050b9f59a0b772d1d0b8c5c1da488a7c36810/packages/api/src/base/Decorate.ts#L347

Check if it is double map and have key1, if false, keep existing behavior. else return prefix with key1 happened.

How the key prefix for double map with a key can be found here

https://github.com/polkadot-js/api/blob/e77050b9f59a0b772d1d0b8c5c1da488a7c36810/packages/api/src/base/Decorate.ts#L466-L472

Update interface

https://github.com/polkadot-js/api/blob/e77050b9f59a0b772d1d0b8c5c1da488a7c36810/packages/api/src/types/storage.ts#L41

Provide appropriate unit tests.

---

@jacogr please tag this if you are happy about this.

Tags:
`bounty-awaiting-approval`
`bounty-S`

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #2213.

### CODE REVIEW

**Comment by KiChjang** [comment]
Unsure where unit tests are located, originally thought `api.spec.ts` would have them, but they do not look like tests.

**Comment by jacogr** [comment]
Thanks a lot for this! Will look in detail a bit later today.

File: packages/api/src/base/Decorate.ts [hunk 1]
```diff
@@ -344,8 +344,15 @@ export default abstract class Decorate<ApiType extends ApiTypes> extends Events
     decorated.key = (arg1?: Arg, arg2?: Arg): string =>
       u8aToHex(compactStripLength(creator(creator.meta.type.isDoubleMap ? [arg1, arg2] : arg1))[1]);
 
-    decorated.keyPrefix = (): string =>
-      u8aToHex(creator.keyPrefix);
+    decorated.keyPrefix = (key1?: Arg): string =>
```
Comment (jacogr) [inline packages/api/src/base/Decorate.ts:347]: The problem here is that we now have duplication of code, as per https://github.com/polkadot-js/api/blob/master/packages/api/src/base/Decorate.ts#L466-L473

So from a maintenance pov we need to ensure that we maintain exactly the same in 2 places. This is problematic.

Comment (KiChjang) [reply packages/api/src/base/Decorate.ts:347]: Can extract this into a function call.

**Comment by KiChjang** [comment]
@jacogr It looks to me that the `iterKey` field on `StorageEntry` is optional, but `keyPrefix` isn't. What exactly is the difference between the two? And should what you said about `iterKey` instead be implemented on `keyPrefix`?

**Comment by jacogr** [comment]
Yes, `iterKey` is optional since it only appears on maps and doublemaps, so in the case of a normal state entry it does not exist. So it really is the key-used-for-iteration.

So what we should do -

- change `iterKey?: Uint8array` to `iterKey?: (arg?: any) => Uint8Array`
- in `kyePrefix` it just becomes `(arg?: any) => iterKey(arg)` (with the conversions to hex as per current)

**Comment by KiChjang** [comment]
@jacogr Ok, I just pushed a new commit up to move the key generation logic into the lower layers, please take a look.

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.ts [hunk 2]
```diff
@@ -157,7 +157,7 @@ function extendHeadMeta (registry: Registry, { meta: { documentation, name, type
     type: registry.createType('StorageEntryTypeLatest', registry.createType('PlainTypeLatest', type.isMap ? type.asMap.key : type.asDoubleMap.key1), 0)
   });
 
-  return registry.createType('StorageKey', iterFn, { method, section });
+  return (arg?: any) => registry.createType('StorageKey', iterFn(arg), { method, section });
```
Comment (xlc) [inline packages/metadata/src/Decorated/storage/fromMetadata/createFunction.ts:160]: One small issue I see here is  when no arguments,`StorageKey` will need to be recreated. Previously it was cached.

```
const prefixKey = registry.createType('StorageKey', iterFn(), { method, section });
return (arg?: any) => !isUndefined(arg) && !isNull(arg)
  ? registry.createType('StorageKey', iterFn(arg), { method, section })
  : prefixKey;
```

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts [hunk 3]
```diff
@@ -104,6 +104,15 @@ describe('createFunction', (): void => {
         (): Uint8Array => storageFn(['5DXUeE5N5LtkW97F2PzqYPyqNkxqSWESdGSPTX6AvkUAhwKP'])
       ).toThrow(/requires two arguments/);
     });
+
+    it.only('accepts an optional parameter for key construction', (): void => {
```
Comment (xlc) [inline packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts:108]:
```suggestion
    it('accepts an optional parameter for key construction', (): void => {
```

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.ts [hunk 4]
```diff
@@ -157,7 +157,11 @@ function extendHeadMeta (registry: Registry, { meta: { documentation, name, type
     type: registry.createType('StorageEntryTypeLatest', registry.createType('PlainTypeLatest', type.isMap ? type.asMap.key : type.asDoubleMap.key1), 0)
   });
 
-  return registry.createType('StorageKey', iterFn, { method, section });
+  const prefixKey = registry.createType('StorageKey', iterFn, { method, section });
+
+  return (arg?: any): StorageKey => !isUndefined(arg) && !isNull(arg)
+    ? registry.createType('StorageKey', iterFn(arg), { method, section })
+    : prefixKey;
```
Comment (jacogr) [inline packages/metadata/src/Decorated/storage/fromMetadata/createFunction.ts:164]:
```suggestion
  return (arg?: any): StorageKey =>
    !isUndefined(arg) && !isNull(arg)
      ? registry.createType('StorageKey', iterFn(arg), { method, section })
      : prefixKey;
```

Just slight OCD, i.e. my eyes follow it better together.

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts [hunk 5]
```diff
@@ -104,6 +104,15 @@ describe('createFunction', (): void => {
         (): Uint8Array => storageFn(['5DXUeE5N5LtkW97F2PzqYPyqNkxqSWESdGSPTX6AvkUAhwKP'])
       ).toThrow(/requires two arguments/);
     });
+
+    it('accepts an optional parameter for key construction', (): void => {
+      expect(storageFn.iterKey).not.toBeUndefined();
+
+      const result = storageFn.iterKey!('5DXUeE5N5LtkW97F2PzqYPyqNkxqSWESdGSPTX6AvkUAhwKP');
```
Comment (jacogr) [inline packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts:111]: No `!` please. (Trying to rid the codebase of them, sadly we have a handful - even in tests, tests are the actual reason we have not been able to switch on the rule, i.e. https://github.com/polkadot-js/api/blob/master/.eslintrc.js#L16)

File: packages/api/src/base/Decorate.ts [hunk 6]
```diff
@@ -455,22 +454,15 @@ export default abstract class Decorate<ApiType extends ApiTypes> extends Events
           .subscribeStorage<[[Codec, Linkage<Codec>]]>([[creator, ...args]])
           .pipe(map(([data]): [Codec, Linkage<Codec>] => data))
         : this._rpcCore.state
-          .subscribeStorage<[LinkageResult]>([creator.iterKey])
+          .subscribeStorage<[LinkageResult]>([creator.keyPrefix()])
```
Comment (jacogr) [inline packages/api/src/base/Decorate.ts:457]:
```suggestion
          .subscribeStorage<[LinkageResult]>([creator.iterKey()])
```

The reason is the `iterKey` is meant to be the iteration key for maps explicitly. (Whereas `keyPrefix` does also exist for non-map types)

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.ts [hunk 7]
```diff
@@ -219,7 +227,7 @@ export default function createFunction (registry: Registry, itemFn: CreateItemFn
     extendDoubleMap(registry, itemFn, storageFn);
   }
 
-  storageFn.keyPrefix = storageFn.iterKey || compactStripLength(storageFn())[1];
+  storageFn.keyPrefix = (arg?: any): Uint8Array => (storageFn.iterKey && storageFn.iterKey(arg)) || compactStripLength(storageFn())[1];
```
Comment (jacogr) [inline packages/metadata/src/Decorated/storage/fromMetadata/createFunction.ts:230]:
```suggestion
  storageFn.keyPrefix = (arg?: any): Uint8Array =>
    (storageFn.iterKey && storageFn.iterKey(arg)) || compactStripLength(storageFn())[1];
```

Small readability change. (This and the previous one is hard to follow, just due to what we are dealing with here)

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts (hunk 5 above)
Comment (KiChjang) [reply packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts:111]: I'm not sure how I can do so, because the TypeScript compiler is not smart enough to know that I've already checked for it to be non-null or non-undefined in a statement above.

File: packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts [hunk 8]
```diff
@@ -4,7 +4,7 @@
 
 import { Text, TypeRegistry } from '@polkadot/types';
 import { StorageEntry } from '@polkadot/types/primitive/StorageKey';
-import { stringToU8a, u8aConcat, u8aToHex } from '@polkadot/util';
+import { stringToU8a, u8aConcat, u8aToHex, assert } from '@polkadot/util';
```
Comment (jacogr) [inline packages/metadata/src/Decorated/storage/fromMetadata/createFunction.spec.ts:7]:
```suggestion
import { assert, stringToU8a, u8aConcat, u8aToHex } from '@polkadot/util';
```
